### DISCUSSION TITLE
Typeface options

### DISCUSSION THREAD

**Discussion by janso3** [ezra-bible-app/ezra-bible-app#1005] [Ideas]
Hi, let me preface this by saying how much I like this app, that's why I would love to have this feature.

Personally, I find the default sans serif typeface not fitting for longer texts, especially biblical ones, since it gives me reading fatigue.
For example on the PC, I like to read the German bible in a Textura/Gothic typeface and, for example, have a different one for Hebrew.

I think even the option of either serif or sans serif would be fine, but I'd love to be able to select any installed typeface.

**Comment by tobias-klein** [comment]
Hi @janso3!

Thanks for your feedback and your idea!

I suppose this should be feasible. I just checked briefly how the list of fonts could be retrieved and I found this nodejs package:
https://github.com/oldj/node-font-list

I suppose this feature would be limited to desktop platforms (Mac, Linux, Windows) or have you also been thinking about mobile (Android)?

Would you apply this only to the Bible text or also to tags, notes, verse references?
Should we consider one setting or is there a use case for having multiple different fonts for different kind of text elements?

**Comment by janso3** [comment]
Hey @tobias-klein,

thank you for your response!
From what I know about Android, it's probably only feasible to include custom font files as resources.
I know of [one app](https://play.google.com/store/apps/details?id=com.skaldebane.fontviewer) that can load them at runtime, but maybe that would complicate things too much.

For my use case at least, I think just changing the look of the Bible view is enough.
I like the way BibleTime and Xiphos let you change fonts per language, but choosing a single font would be plenty for me.

**Comment by tobias-klein** [comment]
Hi @janso3!

I have added issue #1007 now! Let's see whether we can make this happen in the next release?!
Are you interested in contributing for the development?

**Comment by tobias-klein** [comment]
Hi @janso3,

the way how I implemented this now is not yet Bible translation specific. Rather, you can globally change the type face / font and that applies to all Bible translations. Would you be interested in testing the new feature? I could add you as a project member and then you get access to the snapshot builds of the app.

**Reply by janso3** [reply]
Hi,

thank you very much for all this work, this looks very promising. I've been able to build the app for Linux desktop, but couldn't figure out the Android build quite yet. If you have build artifacts for Android I would gladly test those too.

One issue I noticed on my PC is with Hebrew, specifically the OSHB version is missing some glyphs in Serif.
This is not much of a problem since I can just pick a font that has those glyphs, but I wonder if this is also an issue on Android.
![he](https://github.com/ezra-bible-app/ezra-bible-app/assets/25061393/2c1e03c4-6d61-422f-9223-76ad89c6d2f8)

**Reply by tobias-klein** [reply]
Mmmm... the issue with the missing glyphs seems to be system dependent. This is how it looks for me (using Ubuntu 20.04 LTS):

![image](https://github.com/ezra-bible-app/ezra-bible-app/assets/17739895/69f39406-f1b9-44eb-b22c-32ec82e6fcab)

What I noticed though is that I still need to fix the RTL formatting for Hebrew text in the preview box.

**Comment by tobias-klein** [comment]
> thank you very much for all this work, this looks very promising. I've been able to build the app for Linux desktop, but couldn't figure out the Android build quite yet. If you have build artifacts for Android I would gladly test those too.

Thanks for testing! The Android build is slightly complicated ... to say the least. It is done based on the [ezra-bible-app-cordova](https://github.com/ezra-bible-app/ezra-bible-app-cordova) repo. The complicated part is setting up the dependencies correctly. I have to admit that it is not so easy to get started with that independently, sorry. The reason is mainly that nodejs-mobile is a bit dated now and has not been upgraded, because the company that maintained it gave up on it. There is meanwhile an [open source effort](https://github.com/nodejs-mobile/nodejs-mobile) to continue the maintenance based on newer node versions, but the plugin for Cordova has not been uplifted yet.

Having said all that - let me see whether I can supply you a snapshot build for Android.

> One issue I noticed on my PC is with Hebrew, specifically the OSHB version is missing some glyphs in Serif. This is not much of a problem since I can just pick a font that has those glyphs, but I wonder if this is also an issue on Android.

The Serif typeface is literally a CSS style I set correspondingly (`font-family: serif`) and then the webview of Electron / Cordova renders it correspondingly. I don't think there is much we can do to fix that aside from picking a custom font that contains all relevant characters/glyphs.

What could be done is pre-determine which fonts are picked for Hebrew/Greek. This would probably require a per-platform mapping table.

**Comment by tobias-klein** [comment]
@janso3 
I uploaded a current snapshot for Android here:
https://www.dropbox.com/scl/fi/8o4xv1uen33iasnvv0xfd/ezra-bible-app_1.14.0-Dev_2024-02-03.apk?rlkey=cii2r2fdrrin4jlj3vebv4xfn&dl=0

**Comment by tobias-klein** [comment]
@janso3
There may still be some adjustments needed.

Currently the typeface option is only applied to the Bible text but not to the headlines, tags or notes as visible in the following screenshot.

![Screenshot_20240324-072028](https://github.com/ezra-bible-app/ezra-bible-app/assets/17739895/823c73f6-2e1e-46e2-a73a-b0e78bf037fc)

What do you think, should we apply this option also to the headlines, tags and notes?

**Reply by janso3** [reply]
In my opinion, the style should apply to the headlines because they are part of the 'content'.
The tags and notes seem more part of the user interface to me, so sans serif is better.
If I picked, for example, a Gothic typeface, it would seem out of place in the notes.

**Reply by tobias-klein** [reply]
Thanks! Makes sense. I just pushed a change on `master` to also apply the typeface setting for the section titles.

**Reply by tobias-klein** [reply]
@janso3 Just FYI
This change will finally be released as part of version 1.14. I'm planning the release on the next weekend.

**Reply by janso3** [reply]
That is great, thank you very much!
